CI/CD recommendations (#207)
AdobeDocs/graphql-mesh-gateway: src/pages/gateway/cicd.md

---
title: CI/CD for API Mesh
description: Learn how to set up continuous integration and continuous delivery/deployment for API Mesh .
keywords:
  - API Mesh
  - Extensibility
  - GraphQL
  - Integration
  - REST
  - Tools
---

# CI/CD for API Mesh

> API Mesh for Adobe Developer App Builder now provides CI/CD through [Github actions](https://docs.github.com/en/actions). This allows developers to build, test, and deploy meshes faster and more reliably.

Reviewer: We should include a note that explains that API Mesh and App Builder now leverage the same tools to enable CI/CD through Github Actions, so the experience should be consistent for both products.

> ## API Mesh authentication

Author: I've added some numbering. @prabhuram93  Could you clarify what you mean when comparing it to the App Builder docs?

> ## API Mesh authentication

Reviewer 2: Numbering works. I liked the way how this has been split into Lessons. It made it easy to comprehend. It may not apply to this, so just a recommendation to consider the possibility.

<InlineAlert variant="info" slots="text"/>

[Create your mesh](./create-mesh.md) before beginning this process, the supplied GitHub action only handles mesh updates.

To get an OAuth token to enable CI/CD:

> 1. Navigate to the [Adobe Developer Console](https://developer.adobe.com/console) and select your desired workspace.

Reviewer: I think we should should say select the workspace where your mesh is currently, to avoid confusion

    ![Workspace Card](../_images/workspace-card.png)

1. Click the **Add Service** button and select API.

1. From the list of APIs, select the **I/O Management API** under the Adobe Services filter and click **Next**.

    ![IO management API](../_images/io-management-api.png)

1. With OAuth Server-to-Server selected, click **Save configured API**.

1. Click **OAuth Server-to-Server** inside the I/O Management API.

> 1. Click "Generate access token". Copy and save this token for later use.

Reviewer 2: The developers need not generate the token. The GitHub workflow will generate the token based on ClientID, Client secret etc.

Reviewer 2: [This](https://developer.adobe.com/developer-console/docs/guides/authentication/ServerToServerAuthentication/implementation/#generate-access-tokens) is a good article to reference, to generate the access token programmatically. Very useful in the case of BYO CI/CD.

Uncomment the following section to automatically trigger this workflow when a PR is merged `main`:

```yml
push:
  branches: [ "main" ]
```

Uncomment the following section to automatically trigger this workflow when a PR to `main` is updated:

```yml
pull_request:
  branches: [ "main" ]
```

For more information see [Events that trigger workflows](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows).

## Secrets and variables

When developing locally, you can store your variables and secrets in your [environment variables file](./developer-tools.md#environment-variables). When using GitHub, you will need to use [secrets](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions) and [variables](https://docs.github.com/en/actions/learn-github-actions/variables).

- Variables are for information that is not sensitive, for example, if you are targeting `stage` or `prod`. When viewing your mesh in the Developer Console or by running the `aio api-mesh:get` command, these variables will be exposed.

> - Secrets are for sensitive information, such as access tokens. These are never exposed in the Developer Console or when running the `aio api-mesh:get` command.

Reviewer: @revanth0212 is this statement correct in current version?

Reviewer 3: No @naomirobichaud I believe the wording needs to be adjusted to better match our current offering.

There could be confusion here because we are talking about Github Variables and Github Secrets (not API Mesh) and unfortunately we will be using the same naming convention for our future feature **Secrets Management**.

This statement will be valid once we launch our feature later this year.

Meanwhile, I believe we can adjust the wording to only talk about Github Secrets.

```suggestion
- Secrets are for sensitive information, such as access tokens. These are never exposed in the Github code repo.
```

        node-version: ['20']
        os: [ubuntu-latest]
    steps:
      - name: Checkout
        uses: actions/checkout@v4
      - name: Use Node.js ${{ matrix.node-version }}
        uses: actions/setup-node@v4
        with:
          node-version: ${{ matrix.node-version }}
      - name: npm install
        run: npm i
      - name: Setup CLI
        uses: adobe/aio-cli-setup-action@1.3.0
        with:
          os: ${{ matrix.os }}
          version: 10.x.x
      - name: api-mesh-plugin install
        run: aio plugins:install @adobe/aio-cli-plugin-api-mesh
      - name: Auth
        uses: adobe/aio-apps-action@3.3.0
        with:
          os: ${{ matrix.os }}
          command: oauth_sts
          CLIENTID: ${{ secrets.CLIENTID_PROD }}
          CLIENTSECRET: ${{ secrets.CLIENTSECRET_PROD }}
          TECHNICALACCOUNTID: ${{ secrets.TECHNICALACCID_PROD }}
          TECHNICALACCOUNTEMAIL: ${{ secrets.TECHNICALACCEMAIL_PROD }}
          IMSORGID: ${{ secrets.IMSORGID_PROD }}
          SCOPES: AdobeID, openid, read_organizations, additional_info.projectedProductContext, additional_info.roles, adobeio_api, read_client_secret, manage_client_secrets
      - name: Select org

>         run: aio console:org:select 123456

Reviewer 2: This needs to be updated to retrieve from the GitHub secrets. Also applicable for project ID and workspace ID
aio console:org:select ${{ secrets.IMSORGID_PROD }}. 

Should this section be swapped with #4? Since configuring the secrets is a prerequisite to this step.

Reviewer 2: Yes, correct. The workflow templates should not have hardcoded values unless the clients would like to customize them.

          cat .env
      - name: api-mesh update
        run: aio api-mesh:update -c meshConfig.json --env .env 
```

This script sets up the CLI, installs the API Mesh plugin (`aio-cli-plugin-api-mesh`), and adds your OAuth token. Once the workflow is configured, running its corresponding job on the GitHub **Actions** tab will update your mesh.

Uncomment the following section to automatically trigger this workflow when a PR is merged `main`:

```yml
push:
  branches: [ "main" ]
```

Uncomment the following section to automatically trigger this workflow when a PR to `main` is updated:

```yml
pull_request:
  branches: [ "main" ]
```

For more information see [Events that trigger workflows](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows).

## 4, Configure your secrets and variables

When developing locally, you can store your variables and secrets in your [environment variables file](./developer-tools.md#environment-variables). When using GitHub, you will need to use [secrets](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions) and [variables](https://docs.github.com/en/actions/learn-github-actions/variables).

> - Variables are for information that is not sensitive, for example, if you are targeting `stage` or `prod`. When viewing your mesh in the Developer Console or by running the `aio api-mesh:get` command, these variables will be exposed.

Reviewer 3: Same here, I believe we are talking about Github Varilabels which don't have any effect on API Mesh variables unless used in the CICD manually. If that is the case, they will be visible in the mesh. Same is applicable to Secrets as well. This is because API Mesh does not have the concept of Variables and Secrets today. Its all Variables. Once we launch Secrets Management we can call out the distinction in our docs. Meanwhile it would be less confusing if we can limit this section to just Github.

```suggestion
- Variables are for information that is not sensitive, for example, if you are targeting `stage` or `prod` and a variable could carry a different value based on the environment.
```

- [`adobe/aio-apps-action`](https://github.com/adobe/aio-apps-action) centralizes support for a GitHub workflow to leverage application-specific commands, such as deploying via `aio api-mesh:update`.

In the `.github/workflows` folder, create a `.yml` file with the following contents:

```yml
name: CI

# Controls when the workflow will run
on:
  # Triggers the workflow on push or pull request events but only for the "main" branch
  # push:
  #   branches: [ "main" ]
  # pull_request:
  #   branches: [ "main" ]

  # Allows you to run this workflow manually from the Actions tab
  workflow_dispatch:

# A workflow run is made up of one or more jobs that can run sequentially or in parallel
jobs:
  # This workflow contains a single job called "build"
   deploy:
    name: Deploy to Prod
    runs-on: ${{ matrix.os }}
    strategy:
      max-parallel: 1
      matrix:
        node-version: ['20']
        os: [ubuntu-latest]
    steps:
      - name: Checkout
        uses: actions/checkout@v4
      - name: Use Node.js ${{ matrix.node-version }}
        uses: actions/setup-node@v4
        with:
          node-version: ${{ matrix.node-version }}

>       - name: npm install

Reviewer 2: @jhadobe I realized that we do not need an npm install. You can remove this step.

        run: npm i
      - name: Setup CLI
        uses: adobe/aio-cli-setup-action@1.3.0
        with:
          os: ${{ matrix.os }}
          version: 10.x.x
      - name: api-mesh-plugin install
        run: aio plugins:install @adobe/aio-cli-plugin-api-mesh
      - name: Auth
        uses: adobe/aio-apps-action@3.3.0
        with:
          os: ${{ matrix.os }}
          command: oauth_sts
          CLIENTID: ${{ secrets.CLIENTID_PROD }}
          CLIENTSECRET: ${{ secrets.CLIENTSECRET_PROD }}
          TECHNICALACCOUNTID: ${{ secrets.TECHNICALACCID_PROD }}
          TECHNICALACCOUNTEMAIL: ${{ secrets.TECHNICALACCEMAIL_PROD }}
          IMSORGID: ${{ secrets.IMSORGID_PROD }}
          SCOPES: AdobeID, openid, read_organizations, additional_info.projectedProductContext, additional_info.roles, adobeio_api, read_client_secret, manage_client_secrets
      - name: Select org

>         run: aio console:org:select ${{ secrets.ORGID_PROD }}

Reviewer 2: We may need to help differentiate ORGID_PROD and IMSORGID_PROD with examples, as both are not same values. ORGID_PROD is the ID and IMSORGID_PROD is the code.

Author: @prabhuram93 two questions here looking at the download file from dev console: 

1. By `code` do you mean the `"code": "Adobe...",` in the `services` object?

2. Which ID in the following example is the appropriate org id for ORGID_PROD, `id` or `ims_org_id`?

```
  "org": {
      "id": "123456",
      "name": "adobe",
      "ims_org_id": "98765432123456789@AdobeOrg"
  },
```

Reviewer 2: The ORGID_PROD should be "id": "123456"

      - name: Select project
        run: aio console:project:select ${{ secrets.PROJECTID_PROD }}
      - name: Select workspace
        run: aio console:workspace:select ${{ secrets.WORKSPACEID_PROD }}
      - name: 'Create env file'
        run: |
          touch .env
          echo API_ENDPOINT="https://xxx.execute-api.us-west-2.amazonaws.com" >> .env
          echo API_KEY=${{ secrets.API_KEY }} >> .env
          cat .env

>       - name: api-mesh update

Reviewer 2: Since we are mentioning just the update, we should add a note to manually create a mesh on the workspace before setting up the CI/CD. The Expectation is to have Mesh on the workspace on which CI/CD is being set up. 

cc: @naomirobichaud This is a feature gap. We need to do some work to get around this.

Author: @prabhuram93 this is mentioned in a note at the beginning of the page: https://github.com/AdobeDocs/graphql-mesh-gateway/pull/207/files#diff-fc6e4143b29f528c5cc247fea51e30b73af74eb31d7a71e498166362fc3316a8R27

Reviewer 2: Sounds good.

### Add a secret or variable

<InlineAlert variant="info" slots="text"/>

You need administrative permissions to the target GitHub repository to add secrets or variables.

1. Navigate to `https://github.com/<org>/<project_name>/settings/secrets/actions`.

1. On the Secrets or Variables tab, click either **New repository secret** or **New repository variable**.

1. Enter a name for the secret or variable, such as `CLIENTID_PROD`.

1. Enter the corresponding value for the secret or variable, for example, `123456123456`.

The previously described `yml` file in [Create an action](#3-create-a-github-action) requires the following secrets/variables.

```yml
CLIENTID: ${{ secrets.CLIENTID_PROD }}
CLIENTSECRET: ${{ secrets.CLIENTSECRET_PROD }}
TECHNICALACCOUNTID: ${{ secrets.TECHNICALACCID_PROD }}
TECHNICALACCOUNTEMAIL: ${{ secrets.TECHNICALACCEMAIL_PROD }}
IMSORGID: ${{ secrets.IMSORGID_PROD }}
ORGID: ${{ secrets.ORGID_PROD }}
PROJECTID: ${{ secrets.PROJECTID_PROD }}
WORKSPACEID: ${{ secrets.WORKSPACEID_PROD }}
```

For example, if you named a secret `CLIENTID_PROD`, you can call that secret using the `${{ secrets.CLIENTID_PROD }}` format.

> The specific secrets required for this workflow are available in the [Developer Console](https://console.adobe.io/) by navigating to the desired workspace within a project and clicking the **Download all** button in the top-right of the screen. This downloads a `JSON` file.

Reviewer 4: Is `https://console.adobe.io/` the preferred URL? I thought that domain was replaced with `developer.adobe.com`.

- `client_id`
- `client_secrets`
- `technical_account_email`
- `technical_account_id`
- `ims_org_id`

Create a GitHub secret for each of these items.

You also need to specify the `id` for the organization, project, and workspace in the following section of GitHub workflow. These values are also available in the file downloaded from the Developer Console:

```yml
- name: Select org
  run: aio console:org:select 123456
- name: Select project
  run: aio console:project:select 1234567890
- name: Select workspace
  run: aio console:workspace:select 12345678901
```

Repeat this process in your local environment, by adding these values as [environment variables](./developer-tools.md#environment-variables).

This CI/CD process corresponds to a single workspace. If you want to create another workflow, for example, to differentiate between `stage` and `prod`, you need to:

-  Duplicate the workflow you created
-  Replace the workspace `id`

## 4. Create a GitHub action

The GitHub action described in this section updates the specified mesh when a PR is merged to the `main` branch of your repository.

<InlineAlert variant="info" slots="text"/>

> Before you can successfully run this workflow, you must configure your [GitHub secrets](#4-configure-your-secrets-and-variables).

Reviewer 4: Do you need is note? It was done in the previous step. Also, the link should start with 3. Didn't check if the whole string is still correct.

This workflow relies on two existing Adobe I/O GitHub Actions published on the GitHub Marketplace:

> -  `adobe/aio-cli-setup-action` installs and configures the [CLI](https://github.com/adobe/aio-cli) on the GitHub infrastructure. For API Mesh, this action handles authentication.

Reviewer 4: Add a link to these actions?

        run: aio console:project:select ${{ secrets.PROJECTID_PROD }}
      - name: Select workspace
        run: aio console:workspace:select ${{ secrets.WORKSPACEID_PROD }}
      - name: 'Create env file'
        run: |
          touch .env
          echo API_ENDPOINT="https://xxx.execute-api.us-west-2.amazonaws.com" >> .env
          echo API_KEY=${{ secrets.API_KEY }} >> .env
          cat .env
      - name: api-mesh update
        run: aio api-mesh:update -c meshConfig.json --env .env 
```

This script sets up the CLI, installs the API Mesh plugin (`aio-cli-plugin-api-mesh`), and adds your OAuth token. Once the workflow is configured, running its corresponding job on the GitHub **Actions** tab will update your mesh.

Uncomment the following section to automatically trigger this workflow when a PR is merged `main`:

```yml
push:
  branches: [ "main" ]
```

Uncomment the following section to automatically trigger this workflow when a PR to `main` is updated:

```yml
pull_request:
  branches: [ "main" ]
```

For more information see [Events that trigger workflows](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows).

## Bring your own CI/CD

> Although GitHub is the recommended CI/CD platform, it is possible to use other CI/CD platforms. You will need to refer to the CI/CD platform's documentation for secrets management and workflow creation, because the syntax of the workflow and variables will be different for each platform.

Reviewer 4: ```suggestion
Although GitHub is the recommended CI/CD platform, you can use other CI/CD platforms. Refer to the alternative CI/CD platform's documentation for secrets management and workflow creation, because the syntax of the workflow and variables will be different for each platform.
```
